common/formatting/layout_optimizer_internal.h

```diff
@@ -0,0 +1,623 @@
+// Copyright 2017-2021 The Verible Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Implementation details of layout_optimizer.cc exported for tests.
+
+#ifndef VERIBLE_VERILOG_FORMATTING_LAYOUT_OPTIMIZER_INTERNAL_H_
+#define VERIBLE_VERILOG_FORMATTING_LAYOUT_OPTIMIZER_INTERNAL_H_
+
+#include <algorithm>
+#include <iterator>
+#include <ostream>
+#include <type_traits>
+#include <vector>
+
+#include "absl/container/fixed_array.h"
+#include "absl/strings/str_cat.h"
+#include "absl/strings/str_join.h"
+#include "common/formatting/basic_format_style.h"
+#include "common/formatting/token_partition_tree.h"
+#include "common/formatting/unwrapped_line.h"
+#include "common/util/vector_tree.h"
+
+namespace verible {
+
+// LayoutItem type
+enum class LayoutType {
+  // Single line. LayoutItem of this type is always a leaf in LayoutTree.
+  kLine,
+
+  // Joins child items horizontally. See also:
+  // LayoutFunctionFactory::Juxtaposition.
+  kJuxtaposition,
+
+  // Stacks child items vertically. See also: LayoutFunctionFactory::Stack.
+  kStack,
+};
+
+std::ostream& operator<<(std::ostream& stream, LayoutType type);
+
+// LayoutTree node data
+class LayoutItem {
+ public:
+  // Prevent creation of uninitialized LayoutItem
+  LayoutItem() = delete;
+
+  LayoutItem(LayoutType type, int spacing, bool must_wrap, int indentation = 0)
+      : type_(type),
+        indentation_(indentation),
+        spaces_before_(spacing),
+        must_wrap_(must_wrap) {}
+
+  // Creates Line item from UnwrappedLine.
+  explicit LayoutItem(const UnwrappedLine& uwline, int indentation = 0)
+      : type_(LayoutType::kLine),
+        indentation_(indentation),
+        tokens_(uwline.TokensRange()),
+        spaces_before_(!tokens_.empty() ? tokens_.front().before.spaces_required
+                                        : 0),
+        must_wrap_(!tokens_.empty() && tokens_.front().before.break_decision ==
+                                           SpacingOptions::MustWrap) {}
+
+  LayoutItem(const LayoutItem&) = default;
```

Comment:
Shouldn't we rather document the cases where we delete copy, since this is what is often not obvious? Especially in objects that are intended to be data elements - none of which in Verible has copy assign/constructor deleted (except of probably one, which is forced to do so due to use of unique_ptr).
Storing in VectorTree is not the only case where it is used here - we copy it in multiple functions in order to change the copy and then store it in another tree node.

I made `LayoutItem` movable just because it is copyable; to indicate that it is not wrong (although this operation is also required by VectorTree). Move (AFAIK) has to be done explicitly in cases where copy could be intended, so there's no risk of doing it accidentally.

Reply:
Yeah, that is the 'C++ chosing the wrong defaults a lot of the time': We want objects typically non-copyable except when we explicitly consider them data objects. I prefer having the least number of surprises (expensive copies) and then add the functionality where needed, so this is why I prefer delete these and add them when needed. 

Due to the C++ defaults, is easy forgotten and thus we often don't explicitly write it (but we often explicityly make many member variables const for readability, which implicitly prevents a default copy constructor).

Anyway if this is needed in other cases, just adapt the comment to say something like `// used as easy to copy data object` or something.